driver/src/main/scala/api/MongoConnection.scala

```diff
@@ -412,27 +412,48 @@ object MongoConnection {
 
       // ---
 
+      val (unsupportedKeys, options) = opts
       if (useful.indexOf("@") == -1) {
         val (db, hosts) = parseHostsAndDbName(useful)
-        val (unsupportedKeys, options) = opts
 
-        ParsedURI(hosts, options, unsupportedKeys, db, None)
+        options.authMode match {
+          case X509Authentication => ParsedURI(hosts, options, unsupportedKeys, db,
+            Some(Authenticate(db.getOrElse(throw new URIParsingException(s"Could not parse URI '$uri': authentication information found but no database name in URI")), "", "")))
+          case _ => ParsedURI(hosts, options, unsupportedKeys, db, None)
+        }
       } else {
         val WithAuth = """([^:]+):([^@]*)@(.+)""".r
+        val WithAuthNoPassword = """([^:]+)@(.+)""".r
 
         useful match {
           case WithAuth(user, pass, hostsPortsAndDbName) => {
             val (db, hosts) = parseHostsAndDbName(hostsPortsAndDbName)
 
             db.fold[ParsedURI](throw new URIParsingException(s"Could not parse URI '$uri': authentication information found but no database name in URI")) { database =>
-              val (unsupportedKeys, options) = opts
+
+              if (options.authMode == X509Authentication && pass.nonEmpty) {
+                throw new URIParsingException("You should not provide a password when authenticating with X509 authentication")
+              }
 
               ParsedURI(hosts, options, unsupportedKeys, Some(database),
                 Some(Authenticate(options.authenticationDatabase.
                   getOrElse(database), user, pass)))
             }
           }
 
+          case WithAuthNoPassword(user, hostsPortsAndDbName) if options.authMode == X509Authentication => {
+            val (db, hosts) = parseHostsAndDbName(hostsPortsAndDbName)
+
+            db.fold[ParsedURI](throw new URIParsingException(s"Could not parse URI '$uri': authentication information found but no database name in URI")) { database =>
+
+              options.authMode match {
+                case _ => ParsedURI(hosts, options, unsupportedKeys, Some(database),
+                  Some(Authenticate(options.authSource.
+                    getOrElse(database), user, "")))
```

Comment:
`""` seems weird to me

Reply:
@cchantep there is no password, could you please suggest a better alternative